From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Thu, 8 Dec 2022 18:52:05 +0100
Subject: [PATCH] Don't teleport player rider

In the PlayerMoveEvent, the rider Y from the to location is wrong,
but since mojang hasn't exposed a good way
to get the rider offset there's no way to do it without emulating
the positionRider which isn't long term viable (cc camel)
But this patch prevent the teleportation that upstream
do to sync the rider with its vehicle

diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 0c2255b6e2fb7752f85b0f83d4f84732758bd14d..7931c339e8e6b60e2809b446c7c61e8caf24f6e5 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -681,7 +681,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
                 Location curPos = this.getCraftPlayer().getLocation(); // Spigot
 
                 entity.absMoveTo(d3, d4, d5, f, f1);
-                this.player.absMoveTo(d3, d4, d5, this.player.getYRot(), this.player.getXRot()); // CraftBukkit
+                // this.player.absMoveTo(d3, d4, d5, this.player.getYRot(), this.player.getXRot()); // CraftBukkit // Paper don't teleport
 
                 // Paper start - optimise out extra getCubes
                 boolean teleportBack = flag2; // violating this is always a fail
@@ -694,7 +694,7 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
                 }
                 if (teleportBack) { // Paper end - optimise out extra getCubes
                     entity.absMoveTo(d0, d1, d2, f, f1);
-                    this.player.absMoveTo(d0, d1, d2, this.player.getYRot(), this.player.getXRot()); // CraftBukkit
+                    // this.player.absMoveTo(d0, d1, d2, this.player.getYRot(), this.player.getXRot()); // CraftBukkit // Paper don't teleport
                     this.connection.send(new ClientboundMoveVehiclePacket(entity));
                     return;
                 }
@@ -712,12 +712,12 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
                     this.hasMoved = true;
                 }
                 // Spigot End
-                Location from = new Location(player.getWorld(), this.lastPosX, this.lastPosY, this.lastPosZ, this.lastYaw, this.lastPitch); // Get the Players previous Event location.
+                Location from = player.getLocation(); // Paper - Get the actual player location (since their position is updated in the next tick, this become the from loc)
                 Location to = player.getLocation().clone(); // Start off the To location as the Players current location.
 
                 // If the packet contains movement information then we update the To location with the correct XYZ.
                 to.setX(packet.getX());
-                to.setY(packet.getY());
+                to.setY(packet.getY()); // todo wrong rider Y pos
                 to.setZ(packet.getZ());
 
 
